cmake/modules/zephyr_default.cmake

```diff
@@ -0,0 +1,84 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (c) 2021, Nordic Semiconductor ASA
+
+# This CMake module will load all Zephyr CMake modules in correct order for
+# default Zephyr build system.
```

Comment:
> I don't see the need for zephyr_default.cmake.
> 
> Shouldn't modules include their dependencies?

No, cause we have strict module dependencies, for example:
- A module requiring `extensions.cmake` or `generic_toolchain.cmake` because those modules defines functions or configures the toolchain to use.

Or a loose coupling, where one module may set a given set of variables useful for a module, but those variables could be set by other means when reusing the module.
For example, a project such as #40555 could want to reuse `kconfig.cmake` or `dts.cmake` but define it's own set of CMake variables to control the behavior of those modules.

Now, you could argue that in such case, even the `extensions.cmake` or `generic_toolchain.cmake` could be considered loosely coupled as one could simply define all the functions elsewhere.

I have used the following distinction to decide between loose and hard dependency:
- Hard dependency, the module uses a function or macro defined in another module
   - example, modules using functions from `extensions.cmake`
- Hard dependency, the module depends on a program defined / found by another module:
  - Example, `dts.cmake` depends on the C preprocessor found by `generic_toolchain.cmake`
  - `west.cmake` depends on Python, which is fould by `python.cmake` (whoops, just discovered that `include()` in west.cmake is missing)

- Loose dependency, a module depends on variables defined in another module.
  Those variables should be described as an input variable along with a reference to the support module which provides a default way of setting this variable.
  - Example: `boards.cmake` uses `BOARD_ROOT`. The `root.cmake` provides a default way to set `BOARD_ROOT`.
  `kconfig.cmake` uses `CONF_FILE` and `build_configuration.cmake` provides a default way to set this, but when re-using `kconfig.cmake`, such as #40555, one may wish to set it in a different way.
    Therefore thise are loose dependencies.

Reply:
This brings then the reason for the `default_zephyr.cmake`.

It glues everything together for a default Zephyr build, while still allow re-use of individual Zephyr CMake modules independently.
I could place the content of `default_zephyr.cmake` directly here:
https://github.com/zephyrproject-rtos/zephyr/blob/1c57242423cfd3280ed742f44ed33bfd66cbc72d/share/zephyr-package/cmake/ZephyrConfig.cmake#L45

But I felt it was cleaner for Zephyr users to understand what is going on in the default scenario when having a dedicated file under `cmake/modules` than inlined in the Zephyr CMake package config file.

Reply:
This distinction between loose and hard coupling is completely
arbitrary and based on how things have coincidentally been in the
past.

There is no reason to believe users will not want to define their
own function's for something in the future.

There is no reason to believe users will not want to search for
executables with their own business logic.

Does the modules system support providing include paths? If
a user could override which module was used by providing their own
implementation then that could be another way to solve the same problem
that the loose and hard coupling policy is solving.

Reply:
> This distinction between loose and hard coupling is completely
> arbitrary and based on how things have coincidentally been in the
> past.

Not sure I follow.
I tried to explain here: https://github.com/zephyrproject-rtos/zephyr/pull/41301#discussion_r777444109
Let me descxribed the brief rules.

Hard dependency: Things will break, for example missing functions or programs not looked up, fx. C preprocessor or Python.

Loose: General CMake variables. It is now the module's responsibility to ensure sane default values if none is specified. The code in the module should not fail if a setting is not specified, but good values might be provided by other modules.
For example, `zephyr_modules.cmake` may provide extra BOARD_ROOT settings, but running `zephyr_modules.cmake` is not required for running `boards.cmake` or `kconfig.cmake`.

> There is no reason to believe users will not want to define their own function's for something in the future.

and they can still do that any where they like, so i'm unsure what you're asking for.
They can do it in their own toplevel CMakeLists.txt file, or they can create a module of their own and load the functions from there.
If you have a special feature request for something not supported today, then please open a dedicated issue.

> There is no reason to believe users will not want to search for executables with their own business logic.

And they can still do that, just use: [find_program()](https://cmake.org/cmake/help/latest/command/find_program.html)
Not sure what you are asking for here ?
This PR is not limiting what users can do and cannot do.

> Does the modules system support providing include paths? 

I believe I answered that here: https://github.com/zephyrproject-rtos/zephyr/pull/41301#discussion_r777553129
> That would be the CMAKE_MODULE_PATH.

> If a user could override which module was used by providing their own implementation

They can add extra paths to CMAKE_MODULE_PATH and create + load `my_extensions.cmake` in addition to Zephyr CMake modules if they like.

And they can completely control what modules are being loading, for example by calling:
```
find_package(Zephyr COMPONENTS extensions west zephyr_module)
```
as I write here: https://github.com/zephyrproject-rtos/zephyr/pull/41301#issue-805338164

>  the same problem that the loose and hard coupling policy is solving.

Not sure I follow, the loose coupling is not solving anything, because what a loosely coupled module provides should not be required for the other module to work.
I might of course provide extra functionality if available, like Zephyr module or `west` does.

The hard coupling is ensuring that if someone want to run only `dts`, for example like this:
```
find_package(Zephyr COMPONENTS dts)
```
Then that is sufficient to run dts, because the hard deps are automatically loaded.
If we didn't have the hard dependencies, user would need to ensure to write:
```
find_package(Zephyr COMPONENTS extensions python boards generic_toolchain dts)
```
and ensure the CMake modules are loaded in that order.

Reply:
also see a use-case for the loading of selected modules here:
https://github.com/tejlmand/zephyr/blob/05f6a08cb04d2fe5e1526d0bbf8f83820b2273ae/sysbuild/CMakeLists.txt#L11-L18

Reply:
kernel.cmake is an internal file, so it is not necessary for it to include it's dependencies as we can ensure they are included before kernel.cmake is included.